### PR TITLE
feat: reimplement google smart lock for login

### PR DESCRIPTION
**Pull request by addiegupta**
Fixes: #1535

Changes:
Smart Lock features added for Play Store Build Variant
- At Login, an account is chosen. After entering password, Google Smart Lock prompts user to save password for future use
- On next login, account and password data is entered automatically
- In the [previous implementation of this feature](https://github.com/fossasia/open-event-android/pull/990), viewmodel classes had some Android classes being used which is restricted
  Now all classes related to android imports are being handled in LoginFragment and MainActivity, keeping ViewModel android import free
- A SmartAuthUtil class has been made to keep LoginFragment and MainActivity free of Google Play Services imports so that FDroid can be supported
- SmartAuthUtil and SmartAuthViewModel classes for FDroid have also been made which don't do anything
- Build Flavor is being checked before using smart lock features

The need for creating SmartAuthUtil arises as the other alternative is to create 2 files each of LoginFragment and MainActivity for both buildFlavors which might be undesirable

Screenshots for the change:

First Login
smartlocklogin

Second Login: 
smartloginsave
As I have disabled the AutoLogin in my Google Account settings, therefore I need to select 2 more things in the second GIF(first account and then password). This setting is enabled by default for all accounts and for those accounts, email and password will be added automatically.

Additional context:

Implementing these Google APIs has to be done in this scattered manner as it has been designed to used callbacks like `onActivityResult` which are not very suitable for the architecture of this project. I have tried quite a few things but the best solution I could arrive at was this. Please guide me in improving this if you have any suggestions.

### CODE REVIEW

**Comment by addiegupta** [comment]
@iamareebjamal please review

**Comment by iamareebjamal** [comment]
Peer reviews first please

**Comment by addiegupta** [comment]
@liveHarshit @angmas1 please review

**Comment by addiegupta** [comment]
> Peer reviews first please

As you had asked me to work on this, I thought it best to directly ask you for a review as you would have the best insight on this

**Comment by iamareebjamal** [comment]
Peer Reviews teach the reviewing peers on how to review. If they approve and I request some changes, then it means they missed something

File: app/src/main/java/org/fossasia/openevent/general/MainActivity.kt [hunk 1]
```diff
@@ -87,6 +94,23 @@ class MainActivity : AppCompatActivity() {
     fun onSuperBackPressed() {
         super.onBackPressed()
     }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+
+        if (BuildConfig.FLAVOR == "playStore" && requestCode == RC_CREDENTIALS_READ) {
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/MainActivity.kt]: Declare string as constant.

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt [hunk 2]
```diff
@@ -60,6 +64,35 @@ class LoginFragment : Fragment() {
...
+            smartAuthViewModel.password
+                .nonNull()
+                .observe(viewLifecycleOwner, Observer {
+                    password.text = SpannableStringBuilder(it)
+                })
+
+            smartAuthViewModel.resolvableApiException.nonNull().observe(viewLifecycleOwner, Observer {
+                if (smartAuthViewModel.requestCode.value != null) {
+                    SmartAuthUtil.handleResolvableApiException(
+                        it, requireActivity(), smartAuthViewModel.requestCode.value!!)
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: Do not use non-null assertions

File: app/src/main/java/org/fossasia/openevent/general/MainActivity.kt (hunk 1 above)
Comment (addiegupta) [reply app/src/main/java/org/fossasia/openevent/general/MainActivity.kt]: @liveHarshit updated

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt (hunk 2 above)
Comment (addiegupta) [reply app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: In latest commit: As `resolvableApiException` and `requestCode` are always updated together, I have put them together in a `Pair` in the ViewModel and the null case is thus avoided.

**Comment by liveHarshit** [comment]
Build is failing

**Comment by angmas1** [comment]
Run gradlew spotlessApply

File: app/src/main/java/org/fossasia/openevent/general/MainActivity.kt [hunk 3]
```diff
@@ -87,6 +94,23 @@ class MainActivity : AppCompatActivity() {
     fun onSuperBackPressed() {
         super.onBackPressed()
     }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+
+        if (BuildConfig.FLAVOR == getString(R.string.playStore) && requestCode == RC_CREDENTIALS_READ) {
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/MainActivity.kt]: Define as const in the class itself, not in the resources file.

Comment (addiegupta) [reply app/src/main/java/org/fossasia/openevent/general/MainActivity.kt]: updated

**Comment by addiegupta** [comment]
@liveHarshit any other changes that I should make? Or should I proceed with squashing commits?

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt [hunk 4]
```diff
@@ -32,11 +34,15 @@ import org.fossasia.openevent.general.utils.Utils.progressDialog
 import org.fossasia.openevent.general.utils.extensions.nonNull
 import org.jetbrains.anko.design.longSnackbar
 import org.jetbrains.anko.design.snackbar
+import org.koin.androidx.viewmodel.ext.android.sharedViewModel
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
+const val PLAY_STORE_BUILD_FLAVOR = "playStore"
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: Declare it once. It is public value, you can use it in different class also.

Comment (addiegupta) [reply app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: updated. will squash once you approve

**Comment by addiegupta** [comment]
Travis build was failing. removed unnecessary blank line

**Pull request merged**